framework/src/timeintegrators/ActuallyExplicitEuler.C

```diff
@@ -0,0 +1,236 @@
+  const NumericVector<Real> & _diag_inverse;
+};
+
+ActuallyExplicitEuler::ActuallyExplicitEuler(const InputParameters & parameters)
+  : TimeIntegrator(parameters),
+    MeshChangedInterface(parameters),
+    _solve_type(getParam<MooseEnum>("solve_type")),
+    _explicit_residual(_nl.addVector("explicit_residual", false, PARALLEL)),
+    _explicit_euler_update(_nl.addVector("explicit_euler_update", true, PARALLEL)),
+    _mass_matrix_diag(_nl.addVector("mass_matrix_diag", false, PARALLEL))
+{
+  _Ke_time_tag = _fe_problem.getMatrixTagID("TIME");
+
+  // Try to keep MOOSE from doing any nonlinear stuff
+  _fe_problem.solverParams()._type = Moose::ST_LINEAR;
+
+  if (_solve_type == LUMPED || _solve_type == LUMP_PRECONDITIONED)
+    _ones = &_nl.addVector("ones", false, PARALLEL);
+}
+
+void
+ActuallyExplicitEuler::initialSetup()
+{
+  meshChanged();
+}
+
+void
+ActuallyExplicitEuler::init()
+{
+}
+
+void
+ActuallyExplicitEuler::preSolve()
+{
+}
+
+void
+ActuallyExplicitEuler::computeTimeDerivatives()
+{
+  _u_dot = *_solution;
+  _u_dot -= _solution_old;
+  _u_dot *= 1 / _dt;
+  _u_dot.close();
+
+  _du_dot_du = 1.0 / _dt;
+}
+
+void
+ActuallyExplicitEuler::solve()
+{
+  // This ensures that all the Output objects in the OutputWarehouse
+  // have had solveSetup() called, and sets the default solver
+  // parameters for PETSc.
+  _fe_problem.initPetscOutput();
+
+  auto & es = _fe_problem.es();
+
+  auto & nonlinear_system = _fe_problem.getNonlinearSystemBase();
+
+  auto & libmesh_system = dynamic_cast<NonlinearImplicitSystem &>(nonlinear_system.system());
+
+  auto & mass_matrix = *libmesh_system.matrix;
+
+  _current_time = _fe_problem.time();
+
+  // Set time back so that we're evaluating the interior residual at the old time
+  _fe_problem.time() = _fe_problem.timeOld();
+
+  libmesh_system.update();
+
+  // Must compute the residual first
+  _explicit_residual.zero();
+  _fe_problem.computeResidual(*libmesh_system.current_local_solution, _explicit_residual);
+
+  // The residual is on the RHS
+  _explicit_residual *= -1.;
+
+  // Compute the mass matrix
+  _fe_problem.computeJacobianTag(*libmesh_system.current_local_solution, mass_matrix, _Ke_time_tag);
+
+  // Still testing whether leaving the old update is a good idea or not
+  // _explicit_euler_update = 0;
+
+  switch (_solve_type)
+  {
+    case CONSISTENT:
+      _linear_solver->solve(mass_matrix,
+                            _explicit_euler_update,
+                            _explicit_residual,
+                            es.parameters.get<Real>("linear solver tolerance"),
+                            es.parameters.get<unsigned int>("linear solver maximum iterations"));
+
+      break;
+
+    case LUMPED:
+      // Computes the sum of each row (lumping)
+      // Note: This is actually how PETSc does it
+      // It's not "perfectly optimal" - but it will be fast (and universal)
+      mass_matrix.vector_mult(_mass_matrix_diag, *_ones);
```

Comment:
This I do not understand. If you have a `TimeDerivative` kernel with `lumping=true` it will produce a diagonal matrix (notice that the storage in the local matrix is only accessing the diagonal elements). Then, this mat-vec would not be needed. 

I could see that people may not use the lumping parameter but still want to use lumped mass matrix. Then, this would make sense. But then we have 2 ways of doing lumping...

Reply:
I'm going to remove the `lumping` option from `TimeDerivative`.  I'm only leaving it there for now for backward compatibility.

It's better to build the consistent mass matrix - then lump it (for multiple reasons):

 1. With the way I do it you don't need to change anything in your input file to turn on explicit - not ONE single thing.  Just change your time integrator.  This was a huge design goal for this object.
 2. The ability to switch back and forth between explicit and implicit requires that we don't mess with the Kernels themselves (also the same reason why we don't want to use the `implicit = false` options in the input file: note that that's not required for this TimeIntegrator to work)
 3. We often want to build _both_ the lumped mass matrix and consistent mass matrix so that we can use the lumped as a preconditioner.  That's not possible if you lump immediately.
 4. Lumping immediately doesn't save you anything... we still do as much computation and use as much memory in the matrix as when not lumping.  Might as well compute the real (consistent) thing because it gives you more flexibility.

There are a few other small things too.